userland/libtock/alarm.c

```diff
@@ -0,0 +1,220 @@
+  } else {
+    return NULL;
+  }
+}
+
+static void callback( __attribute__ ((unused)) int unused0,
+                      __attribute__ ((unused)) int unused1,
+                      __attribute__ ((unused)) int unused2,
+                      __attribute__ ((unused)) void* ud) {
+  int i = 0;
+  for (alarm_t* alarm = heap_peek(); alarm != NULL; alarm = heap_peek()) {
+    i++;
+    uint32_t now = alarm_read();
+    // has the alarm not expired yet? (distance from `now` has to be larger or
+    // equal to distance from current clock value.
+    if (alarm->expiration - alarm->t0 > now - alarm->t0) {
+      alarm_internal_absolute(alarm->expiration);
+      break;
+    } else {
+      heap_pop(now);
+
+      if (alarm->callback) {
+        // alarm->callback callback _could_ yield, which might result in
+        // reentering this function.
+        tock_enqueue(alarm->callback, now, alarm->expiration, 0, alarm->ud);
+        //alarm->callback(now, alarm->expiration, 0, alarm->ud);
+      }
+      free(alarm);
+    }
+  }
+}
+
+alarm_t *alarm_at(uint32_t expiration, subscribe_cb cb, void* ud) {
+  alarm_t *alarm = (alarm_t*)malloc(sizeof(alarm_t));
+  alarm->t0 = alarm_read();
+  alarm->expiration = expiration;
+  alarm->callback = cb;
+  alarm->ud = ud;
+
+  heap_insert(alarm);
+
+  if (heap_peek() == alarm) {
+    alarm_internal_subscribe((subscribe_cb*)callback, NULL);
+    alarm_internal_absolute(alarm->expiration);
+  }
+
+  return alarm;
+}
+
+alarm_t* alarm_in(uint32_t ms, subscribe_cb cb, void* ud) {
+  uint32_t interval = ms * alarm_internal_frequency() / 1000;
+  uint32_t now = alarm_read();
+  uint32_t expiration = now + interval;
+  return alarm_at(expiration, cb, ud);
+}
+
+struct alarm_repeating {
+  uint32_t interval;
+  subscribe_cb* cb;
+  void* ud;
+  alarm_t* alarm;
+};
+
+static void repeating_cb( uint32_t now,
+                      __attribute__ ((unused)) int unused1,
+                      __attribute__ ((unused)) int unused2,
+                      void* uud) {
+  alarm_repeating_t* udwrapper = (alarm_repeating_t*)uud;
+  uint32_t interval = udwrapper->interval;
+  uint32_t expiration = now + interval;
+  uint32_t cur_exp = udwrapper->alarm->expiration;
+  udwrapper->alarm = alarm_at(expiration, (subscribe_cb*)repeating_cb, (void*)udwrapper);
+  udwrapper->cb(now, cur_exp, 0, udwrapper->ud);
+}
+
+alarm_repeating_t* alarm_every(uint32_t ms, subscribe_cb cb, void* ud) {
+  uint32_t interval = ms * alarm_internal_frequency() / 1000;
+
+  alarm_repeating_t* uud = (alarm_repeating_t*)malloc(sizeof(alarm_repeating_t));
+  uud->interval = interval;
+  uud->cb = cb;
+  uud->ud = ud;
+
+  uint32_t now = alarm_read();
+  uint32_t expiration = now + interval;
+
+  uud->alarm = alarm_at(expiration, (subscribe_cb*)repeating_cb, (void*)uud);
+  return (void*)uud;
+}
+
+void alarm_cancel(alarm_t* alarm) {
+  // Removing from a heap is tricky, so just remove the callback and let it get
+  // lazily removed.
+  alarm->callback = NULL;
```

Comment:
A repeating timer is basically a wrapper around a regular alarm where the callback for the alarm is a shared "repeating_cb" callback that reschedules a new alarm. If the unerlying alarm's callback is set to NULL, then "repeating_cb" won't be called for that timer and it will just fade into oblivion.